### PR TITLE
feat: add color decorator support

### LINKED ISSUE #28: color indicator not working in vue files
in .scss file:
image

in .vue file:
image

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**
Fix #28 . Users can turn decorator off by setting `vue.colorDecorator.enable` to false.

Note #374 requires color picker API, which is still in "proposed" stage now.

Ping @Moerj @aprilandjan 

screen shot 2017-08-21 at 5 30 32 pm

### CODE REVIEW

**Comment by Moerj** [comment]
nice!

**Comment by Moerj** [comment]
Will update in last version?

**Comment by HerringtonDarkholme** [comment]
Ping @octref

File: server/src/vueServerMain.ts [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 import Uri from 'vscode-uri';
 import { DocumentContext, getVls } from './service';
 import * as url from 'url';
 import * as path from 'path';
 import * as _ from 'lodash';
 
+namespace ColorSymbolRequest {
+	export const type: RequestType<string, Range[], any, any> = new RequestType('vue/colorSymbols');
+}
+
```
Comment (octref) [inline server/src/vueServerMain.ts]: You need to reformat everything you get from VSCode repo. They use tab by default.

File: server/src/vueServerMain.ts [hunk 2]
```diff
@@ -178,5 +182,13 @@ connection.onDocumentSymbol(documentSymbolParms => {
...
 });
 
+connection.onRequest(ColorSymbolRequest.type, uri => {
+	const document = documents.get(uri);
+	if (document) {
+		return vls.findColorSymbols(document);
+	}
+	return [];
+});
+
```
Comment (octref) [inline server/src/vueServerMain.ts]: No tabs

File: client/src/colorDecorators.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
+import { window, workspace, DecorationOptions, DecorationRenderOptions, Disposable, Range, TextDocument } from 'vscode';
```
Comment (octref) [inline client/src/colorDecorators.ts]: tabs

File: client/src/htmlMain.ts [hunk 4]
```diff
@@ -37,6 +38,15 @@ export function activate (context: ExtensionContext) {
   const client = new LanguageClient('vue', 'Vue Language Server', serverOptions, clientOptions);
   const disposable = client.start();
   context.subscriptions.push(disposable);
+  const colorRequestor = (uri: string) => {
+    return client.sendRequest(ColorSymbolRequest.type, uri).then(ranges => ranges.map(client.protocol2CodeConverter.asRange));
+  };
+  const isDecoratorEnabled = (languageId: string) => {
+    return workspace.getConfiguration().get<boolean>(languageId + '.colorDecorators.enable');
+  };
```
Comment (octref) [inline client/src/htmlMain.ts]: It seems to be a convention in VSCode extension land to put all setting under your own namesapce, so either `vetur.colorDecorators.enable` (I like this better) or `vetur.vue.colorDecorators.enable`, up to you.

File: package.json [hunk 5]
```diff
@@ -125,6 +125,11 @@
     "configuration": {
       "title": "vetur configuration",
       "properties": {
+        "vue.colorDecorators.enable": {
+          "type": "boolean",
+          "default": true,
+          "description": "Use color decorators in vue"
+        },
```
Comment (octref) [inline package.json]: as mentioned above, everything under `vetur.*`.

**Comment by HerringtonDarkholme** [comment]
Code style updated.

File: client/src/colorDecorators.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  // 0 - 9
+  if (charCode >= 48 && charCode <= 59) {
+    return charCode - 48;
+  }
+  // a - f
+  if (charCode >= 97 && charCode <= 102) {
+    return charCode -87;
+  }
+  if (charCode >= 65 && charCode <= 70) {
+    return charCode -55;
```
Comment (octref) [inline client/src/colorDecorators.ts]: space after `-`. I suggest maybe just run the file through prettier (that's what I do each time).

**Comment by kallebornemark** [comment]
@HerringtonDarkholme Is this feature broken at the moment? Not getting color highlights with `lang="scss"`.

image

**Comment by octref** [comment]
@kallebornemark 

image

Be sure to enable both `"vetur.colorDecorators.enable": true` and `"editor.colorDecorators": true`.

**Pull request merged**